Update to leetcoin application

NoirKomBatman opened the pull request:
#### Project Abstract

This is an amendment to an existing application LeetCoin https://github.com/w3f/Grants-Program/pull/1182.

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [x] **Level 2**:  Up to $30,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] A BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (aUSD/USDT) address for the payment of the milestones is provided in the application.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1550) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1550) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=1550) it.</sub>

NoirKomBatman commented on applications/leetcoin.md:
Previously, we considered S3 to host submission output but realized that Postgres satisfies all our requirements and makes it easier to develop, test locally and maintain in general. The output is stored as JSONB column

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -79,21 +141,37 @@ Legally incorporated in Canada.
-- https://github.com/greenoakxyz/leetcoin-app
+- https://github.com/LeetCoinCo
+- https://github.com/LeetCoinCo/leetcoin-backend
+- https://github.com/LeetCoinCo/leetcoin-frontend
```
We created a separate GitHub org and transferred the repos from greenoakxyz.

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
 ### Milestone 1 — Implement Ink! Problem Builder, Solution and Compiler
 
-- **Estimated duration:** 8 weeks
+- **Estimated duration:** 21 weeks (Completion by March 15, 2023)
```
Unfortunately, due to unforeseen circumstances, we have had to postpone the start of the project to January 2023. We understand that this delay may cause some inconvenience and we apologize for any issues this may cause. The revised completion dates are included in estimate durations

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
+|     3. | Online Judge           | Online judge that compiles Ink! contract code and runs it against test suites comparing the actual output with the expected ones based on the different test cases                                    |  
+|     4. | Frontend problems list | A simple table that displays all available problems on the platform                                                                                                                                   | 
+|     5. | Frontend code editor   | Code editor for the problems                                                                                                                                                                          |  
+|     6. | First Ink! Problem Set | Develop first 5 Ink! problems working with core teams of popular DeFi protocols                                                                                                                       | 
```
After talking with Ink! ecosystem developer, we realized that defining 10 problems is a little difficult and want to consider 5 quality ones first

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
+|     4. | Frontend problems list | A simple table that displays all available problems on the platform                                                                                                                                   | 
+|     5. | Frontend code editor   | Code editor for the problems                                                                                                                                                                          |  
+|     6. | First Ink! Problem Set | Develop first 5 Ink! problems working with core teams of popular DeFi protocols                                                                                                                       | 
+|     7. | Database               | Database to store all problem datasets (Postgres)                                                                                                                                                     | 
```
Updated database storage to Postgres

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
+|    0d. | Article                | Announcement blog post explaining how LeetCoin can be used for upgrading web3 developer skills, improve security and help with properly vetting dev core-team candidates to minimize technical risks. |
+|     1. | Design Document        | Description of architecture and interface                                                                                                                                                             |  
+|     2. | User Authentication    | Create a wallet based authentication system for problem builders as well as users with Web3 Onboard                                                                                                   |  
+|     3. | Online Judge           | Online judge that compiles Ink! contract code and runs it against test suites comparing the actual output with the expected ones based on the different test cases                                    |  
```
`| 3. | Import/Export Data     | Create tools to import and export problem data using multer, csv-express, fast-csv                                                                                                                    | `

is it okay to exclude this one? We realized that since the first problem sets will be hand-crafted with other developers, import/export is not needed and we can populate the problems to our database using SQL. Also, since we use Postgres now as main storage it makes it easier to operate just with SQL.

The problems INSERT statements will be under the `database-seed.sql` or similar seed SQL file so that anyone can have access to them if they run locally.

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
+|     1. | Set Creator                    | Implement problem set curation and creation                                                                                                               |  
+|     2. | Problem Favouriting            | Allow users to favourite specific problems to easily reference later                                                                                      |  
+|     3. | Expand Online Judge            | Allow for problems written in Ink!, Rust and Go by extending compiler                                                                                     |  
+|     4. | Diversify Sets                 | Define and add more problems.                                                                                                                             |  
```
Is it okay to cut down on Educational modules for learners and educators? At this moment we are not sure if they will be beneficial since there are plenty of resources online.

NoirKomBatman commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
-| 2. | Problem Favouriting | Allow users to favourite specific problems to easily reference later                                                                                      |  
-| 3. | Expand Online Judge | Allow for problems written in Ink!, Rust and Go by extending compiler                                                                                     |  
-| 4. | Diversify Sets | Diversify problem sets to include security, testing etc.                                                                                                  |  
-| 5. | Educational Modules (Learner) | A dashboard where learners can explore different popular smart contracts to delve deeply into                                                             | 
```
Is it okay to cut down on Educational modules for learners and educators? At this moment we are not sure if they will be beneficial since there are plenty of resources online.

NoirKomBatman commented:
@keeganquigley @Noc2 hey guys! Thank you for your understanding and sorry for the delay!
Here is an amendment to the original LeetCoin application https://github.com/w3f/Grants-Program/pull/1182

semuelle commented on applications/leetcoin.md:
```diff
@@ -107,81 +185,81 @@ Please also provide the GitHub accounts of all team members. If they contain no
+|     6. | First Ink! Problem Set | Develop first 5 Ink! problems working with core teams of popular DeFi protocols                                                                                                                       | 
+|     7. | Database               | Database to store all problem datasets (Postgres)                                                                                                                                                     | 
+|     8. | API                    | Create an API with express, express router, Postgres and CORS                                                                                                                                         | 
+|    19. | Frontend user profile  | User profile page where they can edit their information and see their progress                                                                                                                        | 
```
```suggestion
|    9. | Frontend user profile  | User profile page where they can edit their information and see their progress                                                                                                                        | 
```

semuelle reviewed (commented):
It all sounds reasonable to me. I found the eduational parts a stretch anyway. Better focus on making the problem sets well. How about, instead of educational material, you add a page with links to educational material? A kind of `Awesome ink!` list?

NoirKomBatman commented:
> It all sounds reasonable to me. I found the eduational parts a stretch anyway. Better focus on making the problem sets well. How about, instead of educational material, you add a page with links to educational material? A kind of `Awesome ink!` list?

Updated: added educational page with awsome Ink! resources

Noc2 reviewed (approved):
Thanks for the update. I will share it with the rest of the team.

semuelle reviewed (approved):
Thanks.
